src/libaktualizr/uptane/ipsecondarydiscovery.cc

```diff
@@ -55,25 +57,32 @@ std::vector<Uptane::SecondaryConfig> IpSecondaryDiscovery::waitDevices() {
 
   while ((recieved = recvfrom(*socket_hdl, rbuf, sizeof(rbuf) - 1, 0, reinterpret_cast<struct sockaddr *>(&sec_address),
                               &sec_addr_len)) != -1) {
-    std::string data(rbuf, recieved);
-    try {
-      Uptane::SecondaryConfig conf;
-
-      int32_t sec_port;
-      asn1::Deserializer des(data);
-
-      des >> asn1::expl(AKT_DISCOVERY_RESP) >> asn1::seq >> asn1::implicit<kAsn1Utf8String>(conf.ecu_serial) >>
-          asn1::implicit<kAsn1Utf8String>(conf.ecu_hardware_id) >> asn1::implicit<kAsn1Integer>(sec_port) >>
-          asn1::restseq >> asn1::endexpl;
-
-      LOG_INFO << "Found secondary:" << conf.ecu_serial << " " << conf.ecu_hardware_id;
-      conf.secondary_type = Uptane::SecondaryType::kIpUptane;
-      conf.ip_addr = sec_address;
-      Utils::setSocketPort(&conf.ip_addr, htons(sec_port));
-      secondaries.push_back(conf);
-    } catch (const deserialization_error &ex) {
-      LOG_ERROR << ex.what();
+    Uptane::SecondaryConfig conf;
+    AKIpUptaneMes_t *m = nullptr;
+    asn_dec_rval_t res = ber_decode(nullptr, &asn_DEF_AKIpUptaneMes, reinterpret_cast<void **>(&m), rbuf, recieved);
+    Asn1Message::Ptr msg = Asn1Message::FromRaw(&m);
```

Comment:
Apparently it can still allocate even if result is an error: "Upon unsuccessful termination, the *struct_ptr_ptr may contain partially decoded
data." From: https://github.com/vlm/asn1c/raw/master/doc/asn1c-usage.pdf

I'll add a note to the source code explain why.

Reply:
... and make `Asn1Message::FromRaw` handle `nullptr`

Reply:
great!